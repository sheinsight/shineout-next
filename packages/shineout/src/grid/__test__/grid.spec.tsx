import { render, cleanup, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Grid from '..';
import {
  baseTest,
  childrenTest,
  classTest,
  // createClassName,
  displayTest,
  snapshotTest,
  styleTest,
} from '../../tests/utils';
import mountTest from '../../tests/mountTest';
import GridBase from '../__example__/01-base';
import GridOffset from '../__example__/02-offset';
import GridNested from '../__example__/03-nested';
import GridGutter from '../__example__/04-gutter';

// const SO_PREFIX = 'grid';
// const originClasses = [''];
// const originItemClasses = ['full', 'md'];
// const { full, md } = createClassName(SO_PREFIX, originClasses, originItemClasses);

const wrapper = '.soui-grid';
const offsetPre = 'soui-offset-md';

const full = 'soui-grid-full';
const md = 'soui-grid-md';

const mdClassName = (f: number | string, s: number | string = '000') => `${md}-${f}-${s}`;
const mdDefaultClassName = mdClassName(100);

const offsetClassName = (f: number | string, s: number | string = '000') =>
  `${offsetPre}-${f}-${s}`;

const stretchStyle = 'min-height: 100%; height: 100%;';

afterEach(cleanup);
mountTest(<Grid />);

describe('Grid[Base]', () => {
  displayTest(Grid, 'ShineoutGrid');
  baseTest(Grid, wrapper);
  childrenTest(Grid, wrapper);
  snapshotTest(<GridBase />);
  snapshotTest(<GridOffset />, 'about offset');
  snapshotTest(<GridNested />, 'about nested');
  snapshotTest(<GridGutter />, 'about gutter');
  test('should render default', () => {
    const { container } = render(<Grid>{'Grid'}</Grid>);
    const gridWrapper = container.querySelector(wrapper)!;
    classTest(gridWrapper, full);
    classTest(gridWrapper, mdDefaultClassName);
  });
  test('should render when set width < 1', () => {
    const width = 1 / 2;
    const widthSecond = 1 / 4;
    const { container, rerender } = render(<Grid width={width}>{'Grid'}</Grid>);
    const gridWrapper = container.querySelector(wrapper)!;
    classTest(gridWrapper, mdClassName(100 * width));
    rerender(<Grid width={widthSecond}>{'Grid'}</Grid>);
    classTest(gridWrapper, mdClassName(100 * widthSecond));
  });
  test('should render when set width > 1', () => {
    const width = 2;
    const { container } = render(<Grid width={width}>{'Grid'}</Grid>);
    const gridWrapper = container.querySelector(wrapper)!;
    classTest(gridWrapper, mdClassName(100 * 1));
  });
  test('should render when set offset', () => {
    const width = 1 / 2;
    const offset = 1 / 4;
    const { container } = render(
      <Grid width={width} offset={offset}>
        {'Grid'}
      </Grid>,
    );
    const gridWrapper = container.querySelector(wrapper)!;
    classTest(gridWrapper, offsetClassName(100 * offset));
  });
  test('should render when set gutter', () => {
    const gutter = 8;
    const gridText = '1 / 8';
    render(
      <Grid gutter={gutter}>
        {Array.from({ length: 8 })
          .map((_, i) => i + 1)
          .map((i) => (
            <Grid key={i} width={1 / 8}>
              {gridText}
            </Grid>
          ))}
      </Grid>,
    );
    const grids = screen.getAllByText(gridText);
    grids.forEach((item) => {
      classTest(item, mdClassName(12, '500'));
      styleTest(item, `padding-left: ${gutter / 2}px; padding-right: ${gutter / 2}px;`);
    });
    styleTest(
      grids[0].parentElement!,
      `width: auto; display: block; margin-left: -${gutter / 2}px; margin-right: -${gutter / 2}px;`,
    );
  });
  test('should render when set stretch', () => {
    const gridText = 'test';
    render(
      <Grid stretch>
        {Array.from({ length: 8 })
          .map((_, i) => i + 1)
          .map((i) => (
            <Grid key={i} width={1 / 8}>
              {gridText}
            </Grid>
          ))}
      </Grid>,
    );
    const grids = screen.getAllByText(gridText);
    grids.forEach((item) => {
      styleTest(item, stretchStyle);
    });
  });
  test('should render when set responsive', () => {
    const responsive = 'sm';
    render(
      <Grid responsive={responsive}>
        {Array.from({ length: 8 })
          .map((_, i) => i + 1)
          .map((i) => (
            <Grid key={i} width={1 / 8}>
              {'test'}
            </Grid>
          ))}
      </Grid>,
    );
    // dont have render
  });
});
