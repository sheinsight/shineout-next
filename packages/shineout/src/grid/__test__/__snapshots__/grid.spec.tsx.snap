// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid[Base] should render correctly  1`] = `
<div>
  <div
    class="cbgmmzo soui-slider-wrapper"
  >
    <div
      class="c-yru59g soui-slider-track"
    >
      <div
        class="c-piwj6u soui-slider-track-inner"
        dir="ltr"
        style="left: 0%; right: 62.5%;"
      >
        <div
          class="c-6bywo0 soui-slider-indicator c6vcnmj soui-slider-indicator-end"
          dir="ltr"
        />
      </div>
    </div>
    <div
      class="c-s6vh6w soui-slider-scale-wrapper"
    >
      <div
        class="c-ysnig5 soui-slider-scale"
      >
        <div
          class="c-ywjbzv soui-slider-label"
          dir="ltr"
        >
          1
        </div>
      </div>
      <div
        class="c-ysnig5 soui-slider-scale"
      >
        <div
          class="c-ywjbzv soui-slider-label"
          dir="ltr"
        >
          2
        </div>
      </div>
      <div
        class="c-ysnig5 soui-slider-scale"
      >
        <div
          class="c-ywjbzv soui-slider-label"
          dir="ltr"
        >
          3
        </div>
      </div>
      <div
        class="c-ysnig5 soui-slider-scale"
      >
        <div
          class="c-ywjbzv soui-slider-label"
          dir="ltr"
        >
          5
        </div>
      </div>
      <div
        class="c-ysnig5 soui-slider-scale"
      >
        <div
          class="c-ywjbzv soui-slider-label"
          dir="ltr"
        >
          8
        </div>
      </div>
      <div
        class="c-ysnig5 soui-slider-scale"
      >
        <div
          class="c-ywjbzv soui-slider-label"
          dir="ltr"
        >
          13
        </div>
      </div>
      <div
        class="c-ysnig5 soui-slider-scale"
      >
        <div
          class="c-ywjbzv soui-slider-label"
          dir="ltr"
        >
          21
        </div>
      </div>
      <div
        class="c-ysnig5 soui-slider-scale"
      >
        <div
          class="c-ywjbzv soui-slider-label"
          dir="ltr"
        >
          34
        </div>
      </div>
      <div
        class="c-ysnig5 soui-slider-scale"
      >
        <div
          class="c-ywjbzv soui-slider-label"
          dir="ltr"
        >
          55
        </div>
      </div>
    </div>
  </div>
  <div
    style="height: 20px;"
  />
  <div
    style="background: rgb(232, 235, 240); margin-bottom: 4px; line-height: 30px;"
  >
    <div
      class="soui-grid soui-grid-full soui-grid-md-20-000 "
      style="color: rgb(255, 255, 255); padding-inline-start: 8px; background: rgb(25, 122, 250); font-size: 14px;"
    >
      1/5
    </div>
  </div>
  <div
    style="background: rgb(232, 235, 240); margin-bottom: 4px; line-height: 30px;"
  >
    <div
      class="soui-grid soui-grid-full soui-grid-md-40-000 "
      style="color: rgb(255, 255, 255); padding-inline-start: 8px; background: rgb(25, 122, 250); font-size: 14px;"
    >
      2/5
    </div>
  </div>
  <div
    style="background: rgb(232, 235, 240); margin-bottom: 4px; line-height: 30px;"
  >
    <div
      class="soui-grid soui-grid-full soui-grid-md-60-000 "
      style="color: rgb(255, 255, 255); padding-inline-start: 8px; background: rgb(25, 122, 250); font-size: 14px;"
    >
      3/5
    </div>
  </div>
  <div
    style="background: rgb(232, 235, 240); margin-bottom: 4px; line-height: 30px;"
  >
    <div
      class="soui-grid soui-grid-full soui-grid-md-80-000 "
      style="color: rgb(255, 255, 255); padding-inline-start: 8px; background: rgb(25, 122, 250); font-size: 14px;"
    >
      4/5
    </div>
  </div>
  <div
    style="background: rgb(232, 235, 240); margin-bottom: 4px; line-height: 30px;"
  >
    <div
      class="soui-grid soui-grid-full soui-grid-md-100-000 "
      style="color: rgb(255, 255, 255); padding-inline-start: 8px; background: rgb(25, 122, 250); font-size: 14px;"
    >
      5/5
    </div>
  </div>
</div>
`;

exports[`Grid[Base] should render correctly about gutter 1`] = `
<div
  style="background: rgb(232, 235, 240);"
>
  <div
    class="soui-grid soui-grid-full soui-grid-md-100-000 "
    style="width: auto; display: block; margin-left: -4px; margin-right: -4px;"
  >
    <div
      class="soui-grid soui-grid-full soui-grid-md-12-500 "
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(25, 122, 250); font-size: 14px;"
      >
        1/8
      </div>
    </div>
    <div
      class="soui-grid soui-grid-full soui-grid-md-12-500 "
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(107, 181, 255); font-size: 14px;"
      >
        1/8
      </div>
    </div>
    <div
      class="soui-grid soui-grid-full soui-grid-md-12-500 "
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(25, 122, 250); font-size: 14px;"
      >
        1/8
      </div>
    </div>
    <div
      class="soui-grid soui-grid-full soui-grid-md-12-500 "
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(107, 181, 255); font-size: 14px;"
      >
        1/8
      </div>
    </div>
    <div
      class="soui-grid soui-grid-full soui-grid-md-12-500 "
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(25, 122, 250); font-size: 14px;"
      >
        1/8
      </div>
    </div>
    <div
      class="soui-grid soui-grid-full soui-grid-md-12-500 "
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(107, 181, 255); font-size: 14px;"
      >
        1/8
      </div>
    </div>
    <div
      class="soui-grid soui-grid-full soui-grid-md-12-500 "
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(25, 122, 250); font-size: 14px;"
      >
        1/8
      </div>
    </div>
    <div
      class="soui-grid soui-grid-full soui-grid-md-12-500 "
      style="padding-left: 4px; padding-right: 4px;"
    >
      <div
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(107, 181, 255); font-size: 14px;"
      >
        1/8
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid[Base] should render correctly about nested 1`] = `
<div
  class="soui-grid soui-grid-full soui-grid-md-100-000 "
  style="text-align: center;"
>
  <div
    class="soui-grid soui-grid-full soui-grid-md-50-000 "
    style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(25, 122, 250); font-size: 14px;"
  >
    1/2
  </div>
  <div
    class="soui-grid soui-grid-full soui-grid-md-50-000 "
    style="line-height: 30px;"
  >
    <div
      style="font-size: 14px;"
    >
      1/2
    </div>
    <div>
      <div
        class="soui-grid soui-grid-full soui-grid-md-33-333 "
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(25, 122, 250); font-size: 14px;"
      >
        1/3
      </div>
      <div
        class="soui-grid soui-grid-full soui-grid-md-33-333 "
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(107, 181, 255); font-size: 14px;"
      >
        1/3
      </div>
      <div
        class="soui-grid soui-grid-full soui-grid-md-33-333 "
        style="color: rgb(255, 255, 255); line-height: 30px; text-align: center; background: rgb(25, 122, 250); font-size: 14px;"
      >
        1/3
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid[Base] should render correctly about offset 1`] = `
<div
  style="background: rgb(232, 235, 240);"
>
  <div
    class="soui-grid soui-grid-full soui-grid-md-33-333 soui-offset-md-33-333"
    style="color: rgb(255, 255, 255); line-height: 60px; text-align: center; background: rgb(25, 122, 250); font-size: 14px;"
  >
    With 1/3, Offset 1/3
  </div>
</div>
`;
